File: DS/bst.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
int data;
struct node *left,*right;
}*root=NULL;

struct node *insert(struct node *root ,int);
void inorder(struct node *root);
void min(struct node *);
void max(struct node *);



struct node *insert(struct node *root ,int data )
{

	if(root==NULL)
	{
	root =(struct node *)malloc(sizeof(struct node)); 
	root->left=NULL;
	root->right=NULL;
	root->data=data;
	printf(" node created %d\n",data);
	return root;
	}
else
{
	 if(data<root->data)
	root->left=insert(root->left,data);
	else if(data > root->data)
	root->right=insert(root->right,data);
	else 
	printf("Same element u enterd\ntry different\n");
	return(root);
}
}
void inorder(struct node *root)
{
if(root!=NULL)
{
inorder(root->left);
printf("%d\n",root->data);
inorder(root->right);
}
}
void postorder(struct node *root)
{
if(root!=NULL)
{
postorder(root->left);
postorder(root->right);
printf("%d\n",root->data);
}
}
void preorder(struct node *root)
{
if(root!=NULL)
{
printf("%d\n",root->data);
preorder(root->left);
preorder(root->right);
}
}
int main()
{
	int choice;
	int i,n,data;
	printf("\n....................BINARY TREES IMPLEMENTATIONS....................................");
	while(1)
	{
		printf("\n\nBinary Search Tree Operations\n");
		printf("\n1. Creation of BST");
		printf("\n2. Traverse in Inorder");
		printf("\n3. Traverse in Preorder");
		printf("\n4. Traverse in Postorder");
		printf("\n5. Find maximum element in the node");
		printf("\n6. Find minumum element in the node");
		printf("\n7. Exit\n");
		printf("\nEnter Choice : ");
		scanf("%d",&choice);
			switch(choice)
			{
				case 1 :
					printf("Enter how many nodes wants to create\n");
					scanf("%d",&n);
					for(i=0;i<n;i++)
					{
						printf("\nEnter data for node %d : ", i);
						scanf("%d",&data);
						root=insert(root,data);
					}
					printf("\nBST with %d nodes is ready to Use!!\n", n);
					break;
				case 2 :
					inorder(root);
					break;
				case 3:
					preorder(root);
					break;
				case 4:
					postorder(root);
					break;
				case 5 :
					printf("The maximum element\n");
					max(root);
					break;
				case 6:
					printf("The minimum element");
					min(root);
					break;
				case 7:
					exit(1);

			}
	}
}

void min(struct node *root)
{
while(root->left !=NULL)
{
root=root->left;
}
printf("%d ",root->data);
}

void  max(struct node *root)
{
while(root->right !=NULL)
{
root=root->right;
}
printf("%d ",root->data);
}

File: PYTHON DS/bst.py
class tree:
    def __init__(self,data=None):
        self.data=data
        if self.data:
            self.left=tree()
            self.right=tree()
        else:
            self.left=None
            self.right=None
        return
    def insert(self,d):
        if self.isempty():
            self.data=d
            self.left=tree()
            self.right=tree()
        if self.data==d:
            return
        if d<self.data:
            self.left.insert(d)
            return
        if d>self.data:
            self.right.insert(d)
            return
    def isleaf(self):
        if self.left==None and self.right==None:
            return True
        else:
            return False
        
    def isempty(self):
        if self.data==None:
            return True
        else:
            return False
        
    def foo(self):
        if self.isempty():
            return(0)
        elif self.isleaf():
            return(self.data)
        else:
            return(self.data+max(self.right.foo(),self.left.foo()))

                
    def inorderTraversal(self, root):
        res = []
        if root:
            res = self.inorderTraversal(root.left)
            res.append(root.data)
            res = res +self.inorderTraversal(root.right)
        return res
root = tree(35)
root.insert(23)
root.insert(46)
root.insert(26)
root.insert(40)
root.insert(39)
root.insert(41)
print(root.foo())
#print(root.inorderTraversal(root))
            
                
                
